### PR TITLE
Practice5 - final solution with mandatory requirements

### PR DESCRIPTION
**Pull request by stratopedarx**

### CODE REVIEW

**Comment by stratopedarx** [comment]
Hello, please review my code.  At the moment, only the mandatory requirements have been implemented.

File: practice_2/WeatherApp/Presenter/WeatherPresenter.swift [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                case .failure(let error):
+                    self.view?.failure(error: error)
+            }
+            group.leave()
+        }
+        group.wait()
+    }
+
+    private func parseWeather(_ weatherApi: WeatherApiModel) {
+        if let coord = weatherApi.coord, let longitude = coord.lon, let latitude = coord.lat,
```
Comment (vasilevkin) [inline practice_2/WeatherApp/Presenter/WeatherPresenter.swift]: put every let ... on the new string.
it's better for readability

File: practice_2/WeatherApp/Services/NetworkService/WeatherNetworkService.swift [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    func fetchForecast(latitude: CLLocationDegrees, longtitude: CLLocationDegrees, units: String,
+                       completion: @escaping (Result<ForecastApiModel, Error>) -> Void) {
+        let params = ["latitude": String(latitude), "longtitude": String(longtitude), "units": units]
+        let request = WeatherApiType.fetchForecast.getRequest(params: params)
+        let task = URLSession(configuration: configuration).dataTask(with: request) { data, _, error in
+            if let error = error {
+                completion(.failure(error))
+                return
+            }
+            if let data = data, let forecast = try? JSONDecoder().decode(ForecastApiModel.self, from: data) {
```
Comment (vasilevkin) [inline practice_2/WeatherApp/Services/NetworkService/WeatherNetworkService.swift]: add a catch to try to handle possible errors during json decoding

File: practice_2/WeatherApp/VIew/Cell/ForecastCell.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    func configure(by forecast: Forecast, unit: String) {
+        if let unixtime = forecast.unixtime, let temperature = forecast.temperature {
+            dateLabel.text = convertUnixtimeToDate(unixtime: unixtime)
+            temperatureLabel.text = "\(String(temperature)) \(unit)"
+        }
+    }
+
+    private func convertUnixtimeToDate (unixtime: Int) -> String {
+        let date = NSDate(timeIntervalSince1970: TimeInterval(unixtime))
+        let formatter = DateFormatter()
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/Cell/ForecastCell.swift]: creating DateFormatter every time is quite heavy and resourceful
 create it once and just reuse if possible

File: practice_2/WeatherApp/VIew/Cell/ForecastCell.swift [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        formatter.dateFormat = "MM-dd-yy EEEE"
+        let dayOfWeek = formatter.string(from: date as Date)
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/Cell/ForecastCell.swift]: handle possible errors in 'as Date' conversion

File: practice_2/WeatherApp/VIew/DetailsWeatherViewController.swift [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import UIKit
+
+class DetailsWeatherViewController: UIViewController {
+    var weather: Weather?
+    var forecastDays = [Forecast?]()
+    var unit: String = defaultUnits.unit
+
+    @IBOutlet private weak var tableView: UITableView!
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/DetailsWeatherViewController.swift:4]: just a minor style remark
@IBOutlet  put it on top of  var 
to have the same code style in the whole project

File: practice_2/WeatherApp/VIew/DetailsWeatherViewController.swift [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            pressureLabel.text = "Pressure: \(info.pressure) in"
+            windSpeedLabel.text = "Wind speed: \(info.windSpeed) m/s"
+            windDegLabel.text = "Wind deg: \(info.windDeg)°"
+            cloudsAllLabel.text = "Cloudness: \(info.cloudsAll)%"
+            weatherImage.image = UIImage(systemName: info.weatherImage)
+        }
+    }
+
+    private func getUnit() -> String {
+        if let unit = UserDefaults.standard.string(forKey: "temperutureUnit") {
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/DetailsWeatherViewController.swift:39]: better to move key to some constant

File: practice_2/WeatherApp/VIew/SettingsViewController.swift [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    @IBOutlet weak var unitSegment: UISegmentedControl!
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setInitialSegmentIndex()
+    }
+
+    private func setInitialSegmentIndex() {
+        if let unit = UserDefaults.standard.string(forKey: "temperutureUnit") {
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/SettingsViewController.swift:18]: better to move key to some constant
using plain strings is error prone potentially

File: practice_2/WeatherApp/VIew/SettingsViewController.swift [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                unitSegment.selectedSegmentIndex = SegmentIndex.imperial.rawValue
+            }
+        } else {
+            return unitSegment.selectedSegmentIndex = SegmentIndex.metric.rawValue
+        }
+    }
+
+    @IBAction func unitChanged(_ sender: UISegmentedControl) {
+        if let unit = sender.titleForSegment(at: sender.selectedSegmentIndex) {
+            UserDefaults.standard.set(unit, forKey: "temperutureUnit")
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/SettingsViewController.swift:31]: the same, key

File: practice_2/WeatherApp/VIew/WeatherViewController.swift [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        self.presenter = WeatherPresenter(view: self, networkService: networkService)
+    }
+
+    private func initRecognizer() {
+        let gestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(tapAnnotationAction))
+        mapView.addGestureRecognizer(gestureRecognizer)
+    }
+
+    private func initSearchController() {
+        let storyboard = UIStoryboard(name: "WeatherApp", bundle: nil)
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/WeatherViewController.swift]: move storyboard names to some constants

File: practice_2/WeatherApp/VIew/WeatherViewController.swift [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            annotation.title = "\(temperature as Double?)"
+        }
+        annotation.coordinate = CLLocationCoordinate2D(latitude: location.latitude, longitude: location.longitude)
+        mapView.addAnnotation(annotation)
+    }
+}
+
+// MARK: MKMapView extension
+extension MKMapView {
+    func centerToLocation(latitude: Double, longitude: Double, regionRadius: CLLocationDistance = 100000) {
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/WeatherViewController.swift]: avoid magic numbers

File: practice_2/WeatherApp/VIew/WeatherViewController.swift [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        let location = mapView.convert(touchLocation, toCoordinateFrom: mapView)
+        return location
+    }
+
+    private func addAnnotation(by location: CLLocationCoordinate2D, temperature: Double?, unit: String) {
+        let annotation = MKPointAnnotation()
+        if let temperature = temperature {
+            annotation.title = "\(temperature) \(unit)"
+        } else {
+            annotation.title = "\(temperature as Double?)"
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/WeatherViewController.swift]: add error handling in as Double conversion

File: practice_2/WeatherApp/VIew/WeatherViewController.swift [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        }
+        self.navigationController?.pushViewController(detailsWeatherVC, animated: true)
+    }
+}
+
+// MARK: WeatherViewProtocol
+extension WeatherViewController: WeatherViewProtocol {
+   func failure(error: Error) {
+        print("FAIL!!!")
+        print(error.localizedDescription)
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/WeatherViewController.swift]: may be it's possible to show some error message to the user in ui,
it's not mandatory, just try to think about possible solution

File: practice_2/WeatherApp/VIew/WeatherViewController.swift [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        // clear existing pins
+        mapView.removeAnnotations(mapView.annotations)
+        let annotation = MKPointAnnotation()
+        annotation.coordinate = placemark.coordinate
+        annotation.title = placemark.name
+        if let city = placemark.locality, let state = placemark.administrativeArea {
+            annotation.subtitle = "\(city) \(state)"
+        }
+        mapView.addAnnotation(annotation)
+        let span = MKCoordinateSpan(latitudeDelta: 0.05, longitudeDelta: 0.05)
```
Comment (vasilevkin) [inline practice_2/WeatherApp/VIew/WeatherViewController.swift]: avoid magic numbers

**Review by vasilevkin — CHANGES_REQUESTED** [review]

File: practice_2/WeatherApp/Presenter/WeatherPresenter.swift (hunk 1 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/Presenter/WeatherPresenter.swift]: thanks. Fix

File: practice_2/WeatherApp/VIew/Cell/ForecastCell.swift (hunk 3 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/VIew/Cell/ForecastCell.swift]: Thanks for that point.
Made formatter as static let.

File: practice_2/WeatherApp/VIew/DetailsWeatherViewController.swift (hunk 5 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/VIew/DetailsWeatherViewController.swift:4]: Done.
Thank you.

File: practice_2/WeatherApp/VIew/WeatherViewController.swift (hunk 9 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/VIew/WeatherViewController.swift]: Fix it.

File: practice_2/WeatherApp/VIew/WeatherViewController.swift (hunk 10 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/VIew/WeatherViewController.swift]: Fixed.
Thanks.

File: practice_2/WeatherApp/VIew/WeatherViewController.swift (hunk 13 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/VIew/WeatherViewController.swift]: Done.

File: practice_2/WeatherApp/Services/NetworkService/WeatherNetworkService.swift (hunk 2 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/Services/NetworkService/WeatherNetworkService.swift]: Try to fix that point. Thanks Sergey.

File: practice_2/WeatherApp/VIew/Cell/ForecastCell.swift (hunk 4 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/VIew/Cell/ForecastCell.swift]: Thanks.
Changed this point in a different way.
Please look into this.

File: practice_2/WeatherApp/VIew/WeatherViewController.swift (hunk 11 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/VIew/WeatherViewController.swift]: Handled that point in another way.
Please have a look. Thanks.

File: practice_2/WeatherApp/VIew/WeatherViewController.swift (hunk 12 above)
Comment (stratopedarx) [reply practice_2/WeatherApp/VIew/WeatherViewController.swift]: Fixed. Add Alert message.
Thanks.

**Comment by stratopedarx** [comment]
Hi @vasilevkin 
Thanks for your review.
Tried to fix all points.
Thanks.

**Review by vasilevkin — APPROVED** [review]
LGTM

**Pull request merged**